File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.java
class Solution {
    public int gcd(int a, int b){
        if(b==0)
            return a;
        else
            return gcd(b, a%b);
    }
    public long gcdSum(int[] nums) {
        int max=-1, n=nums.length;
        int prefixgcd[]=new int[n];
        for(int i=0; i<n; i++){
            max=Math.max(nums[i], max);
            prefixgcd[i]=gcd(nums[i], max);
        }
        long ans=0;
        Arrays.sort(prefixgcd);
        for(int i=0,j=n-1; i<j; i++,j--){
            ans+=gcd(prefixgcd[i], prefixgcd[j]);
        }
        return ans;
    }
}

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.cpp
class Solution {
public:
    int gcd(int a, int b){
        if(b==0)
            return a;
        else
            return gcd(b, a%b);
    }
    long long gcdSum(vector<int>& nums) {
        int maximum=INT_MIN, n=nums.size();
        vector<int> prefixgcd(n);
        for(int i=0; i<n; i++){
            maximum=max(nums[i], maximum);
            prefixgcd[i]=gcd(nums[i], maximum);
        }
        long long ans=0;
        sort(prefixgcd.begin(), prefixgcd.end());
        for(int i=0,j=n-1; i<j; i++,j--){
            ans+=gcd(prefixgcd[i], prefixgcd[j]);
        }
        return ans;
    }
};